src/linear_algebra/multilinear.lean

```diff
@@ -1013,6 +1023,124 @@ def multilinear_curry_right_equiv :
+of multilinear maps on `Π i : ι, M'` taking values in the space of multilinear maps
+on `Π i : ι', M'`. -/
+def curry_sum_equiv : multilinear_map R (λ x : ι ⊕ ι', M') M₂ ≃ₗ[R]
+  multilinear_map R (λ x : ι, M') (multilinear_map R (λ x : ι', M') M₂) :=
+{ to_fun := curry_sum,
+  inv_fun := uncurry_sum,
+  left_inv := λ f, ext $ λ u, by simp,
+  right_inv := λ f, by { ext, simp },
+  map_add' := λ f g, by { ext, refl },
+  map_smul' := λ c f, by { ext, refl } }
+
+variables {ι ι' R M₂ M'}
+
+@[simp] lemma coe_curry_sum_equiv : ⇑(curry_sum_equiv R ι M₂ M' ι') = curry_sum := rfl
+
+@[simp] lemma coe_curr_sum_equiv_symm : ⇑(curry_sum_equiv R ι M₂ M' ι').symm = uncurry_sum := rfl
+
+variables (R M₂ M')
+
+/-- If `s : finset (fin n)` is a finite set of cardinality `k` and its complement has cardinality
+`l`, then the space of multilinear maps on `λ i : fin n, M'` is isomorphic to the space of
+multilinear maps on `λ i : fin k, M'` taking values in the space of multilinear maps
+on `λ i : fin l, M'`. -/
+def curry_fin_finset {k l n : ℕ} {s : finset (fin n)} [decidable_pred (s : set (fin n))]
+  (hk : s.card = k) (hl : sᶜ.card = l) :
+  multilinear_map R (λ x : fin n, M') M₂ ≃ₗ[R]
+    multilinear_map R (λ x : fin k, M') (multilinear_map R (λ x : fin l, M') M₂) :=
+(dom_dom_congr_linear_equiv M' M₂ R R (fin_sum_equiv_of_finset hk hl).symm).trans
+  (curry_sum_equiv R (fin k) M₂ M' (fin l))
+
+variables {R M₂ M'}
+
+lemma curry_fin_finset_apply {k l n : ℕ} {s : finset (fin n)} [decidable_pred (s : set (fin n))]
+  (hk : s.card = k) (hl : sᶜ.card = l) (f : multilinear_map R (λ x : fin n, M') M₂)
+  (mk : fin k → M') (ml : fin l → M') :
+  curry_fin_finset R M₂ M' hk hl f mk ml =
+    f (λ i, sum.elim mk ml ((fin_sum_equiv_of_finset hk hl).symm i)) :=
+rfl
+
+lemma curry_fin_finset_symm_apply {k l n : ℕ} {s : finset (fin n)}
+  [decidable_pred (s : set (fin n))] (hk : s.card = k) (hl : sᶜ.card = l)
+  (f : multilinear_map R (λ x : fin k, M') (multilinear_map R (λ x : fin l, M') M₂))
+  (m : fin n → M') :
+  (curry_fin_finset R M₂ M' hk hl).symm f m =
+    f (λ i, m $ fin_sum_equiv_of_finset hk hl (sum.inl i))
+      (λ i, m $ fin_sum_equiv_of_finset hk hl (sum.inr i)) :=
+rfl
+
+lemma curry_fin_finset_symm_apply_piecewise_diag {k l n : ℕ} {s : finset (fin n)}
+  [decidable_pred (s : set (fin n))] (hk : s.card = k) (hl : sᶜ.card = l)
+  (f : multilinear_map R (λ x : fin k, M') (multilinear_map R (λ x : fin l, M') M₂)) (x y : M') :
+  (curry_fin_finset R M₂ M' hk hl).symm f (s.piecewise (λ _, x) (λ _, y)) = f (λ _, x) (λ _, y) :=
+begin
+  rw curry_fin_finset_symm_apply, congr,
+  { ext i, rw [fin_sum_equiv_of_finset_inl, finset.piecewise_eq_of_mem],
+    apply finset.order_emb_of_fin_mem },
+  { ext i, rw [fin_sum_equiv_of_finset_inr, finset.piecewise_eq_of_not_mem],
+    exact finset.mem_compl.1 (finset.order_emb_of_fin_mem _ _ _) }
+end
+
+lemma curry_fin_finset_symm_apply_diag {k l n : ℕ} {s : finset (fin n)}
```

Comment:
Why `diag` here as a name for `(λ _, x)`?

Reply:
Better suggestions are welcome. I thought about it as the diagonal in `R^n`.

Reply:
For what value of `M'` is this a diagonal in `R^n`? If `M = R^3`, then `(λ _ : fin 3, x)` isn't this a matrix with the same row repeated 3 times?

Reply:
Indeed, it is a counter-intuitive name. Do you have any suggestions? `const` / `piecewise_const`?

Reply:
`const` seems good enough. `same` might be fine too